File: src/main/java/ru/hasa/springbootapp/dto/UserDto.java
package ru.hasa.springbootapp.dto;

import lombok.*;
import ru.hasa.springbootapp.dto.validate.UserCreate;
import ru.hasa.springbootapp.dto.validate.UserUpdate;
import javax.validation.constraints.NotEmpty;
import javax.validation.constraints.NotNull;
import java.util.HashSet;
import java.util.Objects;
import java.util.Set;

@NoArgsConstructor
@AllArgsConstructor
@Getter
@Setter
@Builder
public class UserDto {

    @NotNull(groups = {UserUpdate.class}, message = "Not be empty")
    private Long id;

    @NotEmpty(groups = {UserCreate.class, UserUpdate.class}, message = "Not be empty")
    private String username;

    @NotEmpty(groups = {UserCreate.class, UserUpdate.class}, message = "Not be empty")
    private Byte age;

    @NotEmpty(groups = {UserCreate.class, UserUpdate.class}, message = "Not be empty")
    private String password;

    @NotEmpty(groups = {UserCreate.class, UserUpdate.class}, message = "Not be empty")
    private String mail;

    @NotEmpty(groups = {UserCreate.class, UserUpdate.class}, message = "Not be empty")
    private Set<RoleDto> roles = new HashSet<>();

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        UserDto userDto = (UserDto) o;
        return Objects.equals(id, userDto.id) &&
                Objects.equals(username, userDto.username) &&
                Objects.equals(age, userDto.age) &&
                Objects.equals(password, userDto.password) &&
                Objects.equals(mail, userDto.mail) &&
                Objects.equals(roles, userDto.roles);
    }

    @Override
    public int hashCode() {
        return Objects.hash(id, username, age, password, mail, roles);
    }
}

File: src/main/kotlin/ru/hasa/springbootapp/dto/UserDto.kt
package ru.hasa.springbootapp.dto

import ru.hasa.springbootapp.dto.validate.UserCreate
import ru.hasa.springbootapp.dto.validate.UserUpdate
import javax.validation.constraints.NotEmpty
import javax.validation.constraints.NotNull

data class UserDto(
    @field:NotNull(groups = [UserUpdate::class], message = "Not be empty")
    var id: Long? = null,

    @field:NotEmpty(groups = [UserCreate::class, UserUpdate::class], message = "Not be empty")
    var username: String? = null,

    @field:NotNull(groups = [UserCreate::class, UserUpdate::class], message = "Not be empty")
    var age: Byte? = null,

    @field:NotEmpty(groups = [UserCreate::class, UserUpdate::class], message = "Not be empty")
    var password: String? = null,

    @field:NotEmpty(groups = [UserCreate::class, UserUpdate::class], message = "Not be empty")
    var mail: String? = null,

    @field:NotEmpty(groups = [UserCreate::class, UserUpdate::class], message = "Not be empty")
    var roles: MutableSet<RoleDto> = mutableSetOf()
)
